### DISCUSSION TITLE
Mixing a left panel with a glium render, panel width problems

### DISCUSSION THREAD

**Discussion by prscoelho** [emilk/egui#722] [Q&A] [answered]
I'm running into some weird behavior and I can't help but feel like I'm just doing things wrong. I want to have a egui left panel and then the main view of the app on the right. It's just a bunch of rectangles in 2d. The problem is integrating the panel with the world view, and getting the side panel to play nice with the world view.

Rendering one rectangle at (pane_width, 0),  the quick and dirty way I found to do it was to translate it to the right (by pane_width). This pane_width amount is in world coordinates, so I convert it to window coordinates by `window_width * pane_width / camera_width` so I can tell egui the width size. By printing cursor position and placing it at the top left of the rectangle it seems this value is accurate. The panel however is bigger. Here's how it looks:

![screenshot-2021-09-14 23:40:55+01:00](https://user-images.githubusercontent.com/9609090/133343312-187b0200-7497-42a9-aa43-ee85f79c0a66.png)

The rectangle shouldn't be rendering on the panel at all.

[Here's the code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=54469e25045855cc480c929773994d61), it's essentially just the egui_glium example and the code to render a rectangle.. I guess the meat of the problem is at line 118. Any idea on how to fix this?

**Comment by prscoelho** [comment]
The problem is that I want the panel to be exactly a certain width and it's bigger than that. Setting width to 225, using the mouse to check PhysicalSize and it reports that the panel has 360+ width. Is there extra width being added, or some confusion about what width is in the context of panel width?

**Comment by emilk** [comment, marked as answer]
The best way to figure out what part of the screen that egui isn't covering is to call [`ctx.available_rect()`](https://docs.rs/egui/0.14.2/egui/struct.Context.html#method.available_rect). If you want to render things only in this area you should set the viewport accordingly with whatever graphics API you are using.

The width you can set of a panel is the inner/content width, to which is added the frame.

Not that egui uses logical "points" for all sizes and coordinates, so to get physical pixels you need to multiply with [`pixels_per_point`](https://docs.rs/egui/0.14.2/egui/struct.Context.html#method.pixels_per_point).

**Reply by prscoelho** [reply]
Wow, thank you. You solved every issue I had. Using available_rect I get the remaining space, and using draw paramaters I don't need to translate accounting for the panel, which felt extremely hacky, but I just didn't know any better. The world coordinates shouldn't have to know about the panel at all! Seriously, this is great.

In case anyone ever comes across this same issue, here's the code:

```rust
let available_rect = egui.ctx().available_rect();
let ppp = egui.ctx().pixels_per_point();
let available_width = available_rect.width() * ppp;
let available_height = available_rect.height() * ppp;
let pane_end = window_width - available_width;

let viewport = Some(glium::Rect {
    left: pane_end.ceil() as u32,
    bottom: 0,
    width: available_width.floor() as u32,
    height: available_height.floor() as u32,
});

let params = glium::DrawParameters {
    viewport,
    ..Default::default()
};
```

And here's a screenshot :)

![screenshot-2021-09-23 00:08:40+01:00](https://user-images.githubusercontent.com/9609090/134433699-1c42a342-5c05-4d5a-a2a2-bd8dae87bad9.png)
